datumaro/components/project.py

```diff
@@ -2,486 +2,2383 @@
+                assert osp.isdir(data_dir), data_dir
+                log.debug("Build: loading stage '%s' from '%s'",
+                    current_name, data_dir)
+                return ProjectSourceDataset(
+                    data_dir, self._tree, target, readonly=cached)
+
+            return None
+
+
+        graph = pipeline._graph
+        if len(graph) == 0:
+            raise EmptyPipelineError()
+
+        head = pipeline.head
+        if not head:
+            raise MissingPipelineHeadError()
+
+        # traverse the graph and initialize nodes from sources to the head
+        to_visit = [head]
+        while to_visit:
+            current_name = to_visit.pop()
+            current = graph.nodes[current_name]
+
+            assert current.get('dataset') is None
+
+            obj_hash = current['config'].hash
+            if obj_hash:
+                dataset = _try_load_from_cache(current_name, current)
+                if dataset is not None:
+                    current['dataset'] = dataset
+                    continue
+
+            uninitialized_parents = []
+            initialized_parents = []
+            if graph.in_degree(current_name) == 0:
+                if current['config'].type == 'source':
+                    source = self._tree.sources[
+                        self._tree.build_targets.strip_target_name(current_name)]
+                    if not source.is_generated:
+                        # Source is missing in the cache and cannot
+                        # be retrieved. It is assumed that all the
+                        # sources were downloaded earlier
+                        raise MissingObjectError("Failed to load target '%s': "
+                            "obj '%s' was not found in cache" % \
+                            (current_name, obj_hash))
+                elif current['config'].type == 'project':
+                    pass # pipeline is empty
                 else:
-                    s_config = config.sources[source_name]
-                    if s_config and s_config.format:
-                        # NOTE: consider imported sources as our own dataset
-                        path = None
+                    raise WrongSourceNodeError(
+                        "Unexpected leaf node '%s' of type '%s'" %
+                        (current_name, current['config'].type))
+            else:
+                for p_name in graph.predecessors(current_name):
+                    parent = graph.nodes[p_name]
+                    if parent.get('dataset') is None:
+                        uninitialized_parents.append(p_name)
                     else:
-                        path = [source_name] + (item.path or [])
-                    item = item.wrap(path=path)
+                        initialized_parents.append(p_name)
+
+                if uninitialized_parents:
+                    to_visit.append(current_name)
+                    to_visit.extend(uninitialized_parents)
+                    continue
```

Comment:
This seems like a manual implementation of a topological sort. Wouldn't it be possible to simplify this using `nx.topological_sort`?

Reply:
I feel it can be troublesome, because we need to move in the backward direction from the head node, check if we have data for the node, and move to the previous node only if there is no data available. It is like reversed topological sort with skips (it is describe in the presentations).

Reply:
Okay, I guess that's not really reducible to a topological sort.